### PR TITLE
[FEATURE] Fluent `BatchRequest` slicing

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- DX-471
- Add `batch_slice` to `DataAsset.build_batch_request()` signature
- Update slice parser
- Handle serialization/deserialization in `Checkpoint`
- Add tests for pandas and SQL Fluent Datasources

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/datasource/fluent/sql_datasource.py [hunk 1]
```diff
@@ -666,18 +669,22 @@ def get_batch_list_from_batch_request(
...
-        self, options: Optional[BatchRequestOptions] = None
+        self,
+        options: Optional[BatchRequestOptions] = None,
+        batch_slice: Optional[BatchSlice] = None,
     ) -> BatchRequest:
         """A batch request that can be used to obtain batches for this DataAsset.
 
         Args:
-            options: A dict that can be used to limit the number of batches returned from the asset.
+            options: A dict that can be used to filter the batch groups returned from the asset.
```
Comment (Shinnnyshinshin) [inline great_expectations/datasource/fluent/sql_datasource.py:682]: 👍🏼

File: great_expectations/datasource/fluent/file_path_data_asset.py [hunk 2]
```diff
@@ -41,7 +41,10 @@
     from great_expectations.datasource.fluent.data_asset.data_connector import (
         DataConnector,
     )
-    from great_expectations.datasource.fluent.interfaces import BatchMetadata
+    from great_expectations.datasource.fluent.interfaces import (
+        BatchMetadata,
+        BatchSlice,
```
Comment (Shinnnyshinshin) [inline great_expectations/datasource/fluent/file_path_data_asset.py:45]: is `BatchSlice` an interface that is being introduced in this PR?

File: great_expectations/core/util.py [hunk 3]
```diff
@@ -275,6 +276,7 @@ def convert_to_json_serializable(  # noqa: C901 - complexity 32
             "datasource_name": data.datasource_name,
             "data_asset_name": data.data_asset_name,
             "options": convert_to_json_serializable(data.options),
+            "batch_slice": convert_to_json_serializable(data.batch_slice),
```
Comment (Shinnnyshinshin) [inline great_expectations/core/util.py]: 🚀

File: great_expectations/datasource/fluent/file_path_data_asset.py (hunk 2 above)
Comment (NathanFarmer) [reply great_expectations/datasource/fluent/file_path_data_asset.py:45]: Yes, it is a `TypeAlias` that I added at the top of `batch_filter.py`.

**Review by donaldheppner — CHANGES_REQUESTED** [review]
can you please include an example that loads a batch_request with slices from YAML?

File: tests/datasource/fluent/data_asset/data_connector/test_google_cloud_storage_data_connector.py [hunk 4]
```diff
@@ -72,16 +72,6 @@ def test_basic_instantiation(mock_list_keys):
...
 
-    # Missing "data_asset_name" argument.
-    with pytest.raises(TypeError):
-        # noinspection PyArgumentList
-        my_data_connector.get_batch_definition_list(
-            BatchRequest(
-                datasource_name="something",
-                options={},
-            )
-        )
```
Comment (NathanFarmer) [inline tests/datasource/fluent/data_asset/data_connector/test_google_cloud_storage_data_connector.py:83]: Behavior tested elsewhere

File: tests/datasource/fluent/data_asset/data_connector/test_s3_data_connector.py [hunk 5]
```diff
@@ -79,16 +79,6 @@ def test_basic_instantiation():
...
 
-    # Missing "data_asset_name" argument.
-    with pytest.raises(TypeError):
-        # noinspection PyArgumentList
-        my_data_connector.get_batch_definition_list(
-            BatchRequest(
-                datasource_name="something",
-                options={},
-            )
-        )
```
Comment (NathanFarmer) [inline tests/datasource/fluent/data_asset/data_connector/test_s3_data_connector.py:90]: Behavior tested elsewhere

File: tests/checkpoint/test_checkpoint_result_format.py [hunk 6]
```diff
@@ -300,11 +303,11 @@ def expected_spark_query_output() -> str:
 
 
 def _add_expectations_and_checkpoint(
-    data_context: DataContext | EphemeralDataContext | FileDataContext,
```
Comment (NathanFarmer) [inline tests/checkpoint/test_checkpoint_result_format.py:303]: This wasn't imported, so I'm not sure why we weren't seeing errors.

File: great_expectations/core/util.py [hunk 7]
```diff
@@ -264,18 +266,8 @@ def convert_to_json_serializable(  # noqa: C901 - complexity 32
...
-    from great_expectations.datasource.fluent.interfaces import (
-        BatchRequest as FluentBatchRequest,
-    )
-
-    if isinstance(data, FluentBatchRequest):
-        return {
-            "datasource_name": data.datasource_name,
-            "data_asset_name": data.data_asset_name,
-            "options": convert_to_json_serializable(data.options),
-        }
```
Comment (NathanFarmer) [inline great_expectations/core/util.py:278]: @Kilo59

**Review by donaldheppner — APPROVED** [review]

File: great_expectations/core/util.py [hunk 8]
```diff
@@ -264,18 +266,8 @@ def convert_to_json_serializable(  # noqa: C901 - complexity 32
...
+    if isinstance(data, pydantic.BaseModel):
+        return data.dict()
```
Comment (Kilo59) [inline great_expectations/core/util.py]: @NathanFarmer I think we need to do json round trip to ensure all the nested items are serializable.
```
    if isinstance(data, pydantic.BaseModel):
        return json.loads(data.json())

```

Comment (Kilo59) [reply great_expectations/core/util.py]: Also should update the `ToDict` union to also contain `pydantic.BaseModel`

Comment (NathanFarmer) [reply great_expectations/core/util.py]: Yeah, that makes sense. I forgot that we usually call this recursively, so I could also call `convert_to_json_serializable` on the resulting `dict`,  but I think this is better.

**Pull request merged**